versioned/storage/cache/src/main/java/org/projectnessie/versioned/storage/cache/CaffeineCacheBackend.java

```diff
@@ -48,10 +54,50 @@ static ImmutableCaffeineCacheBackend.Builder builder() {
   @Nullable
   abstract MeterRegistry meterRegistry();
 
+  abstract LongSupplier clockNanos();
+
   @Value.Derived
-  Cache<CacheKey, byte[]> cache() {
-    Caffeine<CacheKey, byte[]> cacheBuilder =
-        Caffeine.newBuilder().maximumWeight(capacity() * 1024L * 1024L).weigher(this::weigher);
+  Cache<CacheKeyValue, byte[]> cache() {
+    Caffeine<CacheKeyValue, byte[]> cacheBuilder =
+        Caffeine.newBuilder()
+            .maximumWeight(capacity() * 1024L * 1024L)
+            .weigher(this::weigher)
+            .expireAfter(
+                new Expiry<CacheKeyValue, byte[]>() {
+                  @Override
+                  public long expireAfterCreate(
+                      CacheKeyValue key, byte[] value, long currentTimeNanos) {
+                    long expire = key.expiresAt;
+                    if (expire == CACHE_UNLIMITED) {
+                      return Long.MAX_VALUE;
+                    }
+                    if (expire == NOT_CACHED) {
+                      return 0L;
+                    }
+                    long currentMicros = NANOSECONDS.toMicros(currentTimeNanos);
```

Comment:
Nit: I think ti would be slightly easier to store nanos directly in `CacheKeyValue.expiresAt`.

Reply:
done